### ISSUE TITLE
loss function question

### ISSUE THREAD

**Issue by zeneofa** [raghavchalapathy/oc-nn#4]
Hi, great idea with the fully deep learning one-class NN. 

I was looking at the codebase and trying to implement a pytorch version. I noticed that the third term of the OC-NN loss function appears different from the paper (line 79 in https://github.com/raghavchalapathy/oc-nn/blob/master/src/models/OC_NN.py, vs equation 4 in https://arxiv.org/pdf/1802.06360.pdf).

In the paper the term uses a dot-product, while the code uses a reduce max. Also in the paper 'w', is said to be the scalar output, though in the code it is the weight of the final layer, so I assume (given that a dot-product is used), 'w' is actually a vector? Also is there a reason for the change from dot-product to reduce max? Am I missing something important?

sorry for all the questions :)

P

**Comment by raghavchalapathy** [comment]
I was experimenting different ways to specify the hinge loss term , with
relu, by using max function all implementation mean the same so you may
choose to use any one which performs better
Thanks

**Comment by germunwang** [comment]
I'm interested in your work. However, as my knowledge, in keras the loss function is for building the connection between the tensors that means it runs just one time in model compile.
As the code in OC_NN.py mentioned by @zeneofa, I think your `self.r` in term3 and term4 will always be `1.0` (since your initial) although you change it in later lines. Maybe you should define `self.r` as a tensorflow variable, then initial and change it also by tensorflow (for example by running a tf.assign).

**Comment by raghavchalapathy** [comment]
I think i have used a tensorflow variable in the code and if you refer to
the new codebase.. you can see the keras callback printing the updated
value of r for every epoch

Raghav

**Comment by germunwang** [comment]
Well I don't know if it is a new codebase, but I just refer to the `OC_NN.py` of your master version in this question, and the issue I said at here:

https://github.com/raghavchalapathy/oc-nn/blob/f57130545f221fee67e9780d2a93ca48b9d10ba5/src/models/OC_NN.py#L72-L92

As I describe as above, keras will run this function at first time only when calling `model.compile()`, instead of `model.fit()`. It will not run in every epoch while fitting.

So that even though you update `self.r` in L82 and L84, the `r` value for `term3` and `term4` in L79-80 will not be updated for next epoch, since it will be a constant tensor of value `1.0` by your initial define in: 
https://github.com/raghavchalapathy/oc-nn/blob/f57130545f221fee67e9780d2a93ca48b9d10ba5/src/models/OC_NN.py#L63

So if you only callback the `self.r` in L84, then it may show the truly updated `r` value.
However, if you callback the value of `term4`, you will always get `-1.0` every epoch, and so is `r=1.0` in term3.

Maybe you could just callback the `term4` then you will understand what I mean,
or try callback `term3` comparing with following:
`check_term3 = 1 / nu * K.mean(K.maximum(0.0, 1.0 - tf.reduce_max(y_pred, axis=1)), axis=-1)`
then you will find out this is equal to `term3` even though you update `self.r` in every epoch.

**Comment by raghavchalapathy** [comment]
Hello

Many thanks for your email and interest  in the OC-NN work
Could you please see the example of OC-NN here
<https://github.com/raghavchalapathy/oc-nn/blob/f57130545f221fee67e9780d2a93ca48b9d10ba5/notebooks/OC-NN/MNIST/OCNN_Experiments.ipynb>
as
pointed out in the example presented in the github readme pages

Kindly follow the OC-NN implementation within the file

from src.models.OneClass_SVDD import OneClass_SVDD


OC_NN.py is the legacy code you were referring to, my bad should have
removed the file to avoid confusion.

Also, I am happy to do a code review together to see if there is any bug in
the implementation in the file OneClass_SVDD.py
To the best of my knowledge, updating the r value is handled correctly
after each epoch.

But yes I am happy to discuss if you have any concerns

with regards
Raghav

**Comment by raghavchalapathy** [comment]
Closing due to inactivity
